File: kotlin/rail-fence-cipher/src/main/kotlin/RailFenceCipher.kt
class RailFenceCipher(private val rails: Int) {

    fun getEncryptedData(input: String): String {
        if (rails == 0) return input

        val rows = Array(rails) { StringBuilder() }

        var rail = 0
        var direction = 1

        for (c in input) {
            rows[rail].append(c)
            if (rail == 0) {
                direction = 1
            } else if (rail == rails - 1) {
                direction = -1
            }

            rail += direction
        }

        return rows.joinToString("") { it.toString() }
    }

    fun getDecryptedData(input: String): String {
        if (rails == 1) return input

        val pattern = IntArray(input.length)

        var rail = 0
        var direction = 1

        for (i in input.indices) {
            pattern[i] = rail

            if (rail == 0) {
                direction = 1
            } else if (rail == rails - 1) {
                direction = -1
            }

            rail += direction
        }

        val counts = IntArray(rails)
        for (r in pattern) {
            counts[r]++
        }

        val railChars = Array(rails) { CharArray(counts[it]) }
        var index = 0

        for (r in 0 until rails) {
            for (i in railChars[r].indices) {
                railChars[r][i] = input[index++]
            }
        }

        val railIndices = IntArray(rails)
        return buildString {
            for (r in pattern) {
                append(railChars[r][railIndices[r]++])
            }
        }
    }
}

File: csharp/rail-fence-cipher/RailFenceCipher.cs
using System.Text;

public class RailFenceCipher(int rails)
{
    private readonly int rails = rails;

    public string Encode(string input)
    {
        if (rails == 0) return input;

        var rows = new string[rails];

        var rail = 0;
        var direction = 1;

        foreach (var c in input)
        {
            rows[rail] += c;

            if (rail == 0)
            {
                direction = 1;
            }
            else if (rail == rails - 1)
            {
                direction = -1;
            }

            rail += direction;
        }

        return string.Join("", rows);
    }

    public string Decode(string input)
    {
        if (rails == 1) return input;

        var pattern = new int[input.Length];
        var rail = 0;
        var direction = 1;

        for (var i = 0; i < input.Length; i++)
        {
            pattern[i] = rail;

            if (rail == 0)
            {
                direction = 1;
            }
            else if (rail == rails - 1)
            {
                direction = -1;
            }

            rail += direction;
        }

        var counts = new int[rails];

        foreach (var r in pattern)
        {
            counts[r]++;
        }

        var railChars = new char[rails][];

        for (var i = 0; i < rails; i++)
        {
            railChars[i] = new char[counts[i]];
        }

        var index = 0;
        for (var r = 0; r < rails; r++)
        {
            for (var i = 0; i < railChars[r].Length; i++)
            {
                railChars[r][i] = input[index++];
            }
        }

        var railIndicies = new int[rails];
        var builder = new StringBuilder();

        foreach (var r in pattern)
        {
            builder.Append(railChars[r][railIndicies[r]++]);
        }

        return builder.ToString();
    }
}